code/game/machinery/tcomms/core.dm

```diff
@@ -121,6 +128,38 @@
 			reachable_zlevels |= R.loc.z
 
 
+/**
+  * Z-Level transit change helper
+  *
+  * Handles parent call of disabling the machine if it changes Z-level, but also rebuilds the list of reachable levels
+  */
+/obj/machinery/tcomms/core/onTransitZ(old_z, new_z)
+	. = ..()
+	refresh_zlevels()
+
+/**
+  * Power-on checker
+  *
+  * Checks the z-level to see if an existing core is already powered on, and deny this one turning on if there is one. Returns TRUE if it can power on, or FALSE if it cannot
+  */
+/obj/machinery/tcomms/core/proc/check_power_on()
```

Comment:
Wouldn't it be better to declare this at tcomms level instead of the core and relay separately?
They both are nearly the same yet search for different machines. Best to avoid dupe code

Reply:
I defined it at this level because of the typepath change in the `GLOB.tcomms_machines` list, and theres so much of a change between those that this made more sense ot me

Reply:
You can make a proc to return the list type. Then loop through it with a `var/thing` and cast it to `obj/machinery/tcomms`. I don't think you have any actual checks that relate to core or relay right? If so you can make a proc for the checks

Reply:
I do, the checks need to be separate between a core or a relay, so you can have 1 core AND 1 relay. Not 1 core OR relay.

Reply:
You don't have to check variables on the core or relay though. Only on core level.
My idea was like this
```
.../tcomms/proc/get_machines_to_check()
  return
...

...tcomms/proc/check_power_on()
  ..
  var/list/machines = get_machines_to_check() 
  for(var/thing in machines)
    var/.../tcomms/TC = thing
    ....

.../core/core/get_machines_to_check()
  var/list/result = list()
  for(var/obj/machinery/tcomms/core/C in GLOB.tcomms_machines)
    result += C
  return result
...
```
This way you reduce dupe code and make it OO

Reply:
Defining a blank proc just to override it in 2 places seems bad in my eyes, so I ain't doing it.

Reply:
That is... how OO works though... look at attackby. attack etc. All pretty empty until behaviour is defined. Or wrench_act etc

Reply:
Yeah but the difference is that has about 1000 re-definitions which all rely on parent calls. These **two** procs are written in very specific ways to iterate through the loop and only get the specific machine.